include/GafferScene/Shader.h

```diff
@@ -121,6 +121,9 @@ class Shader : public Gaffer::DependencyNode
 		/// \deprecated
 		IECore::ConstObjectVectorPtr state() const;
 
+		virtual void attributesHash( IECore::MurmurHash &h, const Gaffer::Plug *output ) const;
```

Comment:
I don't think that will work, because NetworkBuilder needs access to the protected `Shader::parameterHash()` and `Shader::parameterValue()` methods anyway. I think we should just make ShaderPlug a friend of Shader. Let me explain...

The other gaffer process types are organised with a pairing of a plug type and a node type. So we have ComputeNode/ValuePlug and TaskNode/TaskPlug for instance. In each case, the protected node internals are the workhorses that are responsible for getting things done (`hash()` and `compute()` for ComputeNode and `hash()` and `execute()` for TaskNode). And in each case, the plugs are the public interface, using friend access to provide some insulation from the internals, and removing some hazards for clients :

- All plugs implicitly use the current context in `getValue()` or `execute()` so the user can't get the contexts mismatched, but the node methods pass this around internally to avoid lots of thread-local lookups
- In the case of promoted output plugs and other chained connections, plugs automatically trace back to the source to find the right node to call `compute()` or `execute()` on
- ValuePlug introduces a caching layer

I've been gradually refactoring Shader/ShaderPlug to match this pattern, and at some point also need to bring `DependencyNode::affects()` into the fold (it's still public). That's why I was hoping you could continue in this direction by making the protected virtual methods, and keeping NetworkBuilder as the internal "workhorse" of the node. Do you see any reason why that can't work?